File: libraries/rust/src/services.rs
use crate::identity::{EndpointAddress, IdentityError};

pub const SERVICE_COMMAND: &str = "serviceCommand";
pub const SERVICE_COMMAND_REPLY: &str = "serviceCommandReply";

pub fn service_address(service_id: &str) -> Result<EndpointAddress, IdentityError> {
    EndpointAddress::new("service", service_id)
}

pub use crate::keys::parse_service_catalog_key as parse_catalog_key;
pub use crate::keys::parse_service_status_key as parse_status_key;
pub use crate::keys::parse_service_view_key as parse_view_key;
pub use crate::keys::service_catalog_key as catalog_key;
pub use crate::keys::service_status_key as status_key;
pub use crate::keys::service_view_key as view_key;

File: libraries/typescript/src/services.ts
import type { DeckrMessage, EntitySubject } from "./identity.js";
import { serviceAddress } from "./identity.js";
import {
  parseServiceCatalogKey,
  parseServiceStatusKey,
  parseServiceViewKey,
  serviceCatalogKey,
  serviceStatusKey,
  serviceViewKey,
  type JsonObject,
  type ServiceCatalog,
  type ServiceStatus,
} from "./state.js";

export const SERVICE_COMMAND = "serviceCommand";
export const SERVICE_COMMAND_REPLY = "serviceCommandReply";

export {
  parseServiceCatalogKey,
  parseServiceStatusKey,
  parseServiceViewKey,
  serviceAddress,
  serviceCatalogKey,
  serviceStatusKey,
  serviceViewKey,
  type JsonObject,
  type ServiceCatalog,
  type ServiceStatus,
};

export function serviceBody(message: DeckrMessage): JsonObject {
  if (message.messageType !== SERVICE_COMMAND && message.messageType !== SERVICE_COMMAND_REPLY) {
    throw new Error(`unsupported service message type ${message.messageType}`);
  }
  return message.body;
}

export function serviceSubject(input: {
  serviceId: string;
  namespace: string;
  operation: string;
}): EntitySubject {
  return {
    kind: "service",
    identifiers: {
      namespace: input.namespace,
      operation: input.operation,
      serviceId: input.serviceId,
    },
  };
}
